lib/clustering/ocn_graph.rb

```diff
@@ -0,0 +1,73 @@
+# frozen_string_literal: true
+
+require "cluster"
+
+module Clustering
+
+  # Constructs a graph from OCN tuples
+  class OCNGraph
+
+    attr_accessor :edges, :nodes
+
+    def initialize(cluster = nil)
+      @cluster = cluster
+      @nodes = Set.new
+      # root_ocn : [all matching ocns]
+      @edges = Hash.new {|g, root| g[root] = Set.new }
+      compile_graph_from_cluster unless @cluster.nil?
+    end
+
+    # Add the tuple of OCNs to the list of nodes and edges
+    #
+    # @param [Enumerable] tuple A set of one or more OCNs from a clusterable
+    def add_tuple(tuple)
+      nodes.merge(tuple)
+      tuple.map {|node| edges[node] << node }
+      tuple.sort.to_a.combination(2).each do |parent, child|
+        edges[parent] << child
+        edges[child] << parent
+      end
+    end
+
+    # root, children = edges.sort.first
+
+    # Partition the :nodes into one or more subgraphs
+    #
+    # TODO: better name. really just subsets of OCNS
+    def subgraphs
```

Comment:
Is there an opportunity for extracting any graph-related functionality from https://github.com/hathitrust/holdings-backend/blob/main/lib/concordance_validation/concordance.rb#L69 ? It may also be worth taking a look at https://github.com/monora/rgl. It looks like `add_tuple` will result in strongly-connected components, so https://ruby-doc.org/stdlib-3.0.1/libdoc/tsort/rdoc/TSort.html might also work.

Reply:
Concordance validation needs some attention on its own, if only because we don't actually do anything about the warnings it generates. It's also a directed graph making it a candidate for TSort while our clustering is not. RGL looks like a promising replacement for OCNGraph

Reply:
"directed graph" - true, but you can model an undirected graph as a directed graph with two edges for each pair of vertices connected in the undirected graph, and I believe the notion of strongly connected component on such a directed graph corresponds to the notion of connected component in the undirected graph. I had thought we were already doing that because of

```ruby
        edges[parent] << child
        edges[child] << parent
```